modules/codegen/src/smithy4s/codegen/ModelLoader.scala

```diff
@@ -25,82 +25,67 @@ import software.amazon.smithy.model.Model
 import software.amazon.smithy.model.loader.ModelAssembler
 
 import java.io.File
-import java.net.URL
 import java.net.URLClassLoader
 
 import scala.jdk.CollectionConverters._
+import software.amazon.smithy.model.loader.ModelDiscovery
+import software.amazon.smithy.model.loader.ModelManifestException
 
 object ModelLoader {
 
   def load(
-      specs: Set[File],
+      localSpecs: Set[File],
       dependencies: List[String],
       repositories: List[String],
       transformers: List[String],
       discoverModels: Boolean,
       localJars: List[os.Path]
   ): (ClassLoader, Model) = {
-    val allDeps = dependencies
-    val maybeDeps = resolveDependencies(allDeps, localJars, repositories)
     val currentClassLoader = this.getClass().getClassLoader()
+    val deps = resolveDependencies(dependencies, localJars, repositories)
+
+    val modelsInJars = deps.flatMap { files =>
+      val manifestUrl =
+        ModelDiscovery.createSmithyJarManifestUrl(files.getAbsolutePath())
+      try { ModelDiscovery.findModels(manifestUrl).asScala }
+      catch {
+        case _: ModelManifestException => Seq.empty
+      }
+    }
 
-    // Loads a model using whatever's on the current classpath (in particular, anything that
-    // might be provided by smithy4s itself out of the box)
-    val modelBuilder = Model
+    // Loading the upstream model
+    val upstreamModel = Model
       .assembler()
+      // disabling cache to support snapshot-driven experimentation
+      .putProperty(ModelAssembler.DISABLE_JAR_CACHE, true)
       .addClasspathModels(currentClassLoader, discoverModels)
+      .addImports(modelsInJars)
       .assemble()
       .unwrap()
-      .toBuilder()
```

Comment:
this logic is made simpler by the fact that we're removing the assumption that (alloy) models are communicated through the classpath. The plugins (SBT, mill, CLI) are now responsible for setting the dependency to alloy as default.